homeassistant/components/sia/alarm_control_panel.py

```diff
@@ -0,0 +1,249 @@
+
+    async def async_added_to_hass(self) -> None:
+        """Run when entity about to be added to hass.
+
+        Overridden from Entity.
+
+        1. start the event listener and add the callback to on_remove
+        2. get previous state from storage
+        3. if previous state: restore
+        4. if previous state is unavailable: set _available to False and return
+        5. if available: create availability tracker
+        """
+        self.async_on_remove(
+            self.hass.bus.async_listen(
+                event_type=SIA_EVENT.format(self._port, self._account),
+                listener=self.async_handle_event,
+            )
+        )
+        last_state = await self.async_get_last_state()
+        if last_state is not None:
+            self.state = last_state.state
+        if self.state == STATE_UNAVAILABLE:
+            self._available = False
+            return
+        self._cancel_availability_tracker = self.async_create_availability_tracker()
+
+    async def async_will_remove_from_hass(self) -> None:
+        """Run when entity will be removed from hass.
+
+        Overridden from Entity.
+        """
+        if self._cancel_availability_tracker:
+            self._cancel_availability_tracker()
+
+    async def async_handle_event(self, event: Event) -> None:
+        """Listen to events for this port and account and update states.
+
+        If the port and account combo receives any message it means it is online and can therefore be set to available.
+        """
+        sia_event = SIAEvent.from_dict(event.data)  # pylint: disable=no-member
+        _LOGGER.debug("Received event: %s", sia_event)
+        if int(sia_event.ri) == self._zone:
+            self.state = CODE_CONSEQUENCES.get(sia_event.code, None)
+            self._attr.update(get_attr_from_sia_event(sia_event))
+        self._available = True
+        self.async_write_ha_state()
+        self.async_reset_availability_tracker()
+
+    @callback
+    def async_reset_availability_tracker(self) -> None:
+        """Reset availability tracker by cancelling the current and creating a new one."""
+        if self._cancel_availability_tracker:
+            self._cancel_availability_tracker()
+        self._cancel_availability_tracker = self.async_create_availability_tracker()
+
+    @callback
+    def async_create_availability_tracker(self) -> CALLBACK_TYPE:
+        """Create a availability tracker and return the callback."""
+        return async_call_later(
+            self.hass,
+            get_unavailability_interval(self._ping_interval),
+            self.async_set_unavailable,
+        )
+
+    @callback
+    def async_set_unavailable(self, _) -> None:
+        """Set unavailable."""
+        self._available = False
+        self.async_write_ha_state()
+
+    @property
+    def state(self) -> StateType:
+        """Get state."""
+        return self._state
+
+    @state.setter
```

Comment:
Second opinion needed on if our standards allow a setter on an entity when its not being called from outside the entity.

Reply:
you're absolutely right, don't need the state.setter since there is only one place where that logic happens, so moved it there and got rid of the setter!